File: .github/scripts/check_local_links.py
#!/usr/bin/env python3
"""Check that local href/src references in HTML files resolve to real files."""
import sys
from html.parser import HTMLParser
from pathlib import Path
from urllib.parse import urlsplit

REPO_ROOT = Path(__file__).resolve().parents[2]
ATTRS_TO_CHECK = {"src", "href"}
SKIP_PREFIXES = ("http://", "https://", "//", "mailto:", "tel:", "javascript:", "data:")


class LinkExtractor(HTMLParser):
    def __init__(self):
        super().__init__()
        self.refs = []

    def handle_starttag(self, tag, attrs):
        for name, value in attrs:
            if name in ATTRS_TO_CHECK and value:
                self.refs.append(value)


def local_path_for(ref, html_file):
    if ref.startswith(SKIP_PREFIXES) or ref.startswith("#"):
        return None
    path = urlsplit(ref).path
    if not path:
        return None
    if path.startswith("/"):
        return REPO_ROOT / path.lstrip("/")
    return html_file.parent / path


def main():
    html_files = sorted(REPO_ROOT.glob("*.html"))
    missing = []
    for html_file in html_files:
        parser = LinkExtractor()
        parser.feed(html_file.read_text(encoding="utf-8", errors="replace"))
        for ref in parser.refs:
            target = local_path_for(ref, html_file)
            if target is not None and not target.exists():
                missing.append(f"{html_file.relative_to(REPO_ROOT)}: missing '{ref}'")

    if missing:
        print("Broken local references found:")
        for line in missing:
            print(f"  - {line}")
        return 1

    print(f"All local references in {len(html_files)} HTML file(s) resolve correctly.")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/check-local-links.js
#!/usr/bin/env node
// Verifies that every local asset reference (href/src) in the site's HTML
// files points to a file that actually exists in the repository.
'use strict';

const fs = require('fs');
const path = require('path');

const ROOT = path.resolve(__dirname, '..');
const HTML_FILES = ['index.html', '404.html'];
const ATTR_PATTERN = /\b(?:href|src)\s*=\s*["']([^"']+)["']/g;

function isLocal(ref) {
  if (!ref || ref.startsWith('#')) return false;
  if (/^([a-z][a-z0-9+.-]*:)/i.test(ref)) return false; // has a scheme (http:, mailto:, tel:, etc.)
  if (ref.startsWith('//')) return false; // protocol-relative
  return true;
}

let missing = [];

for (const file of HTML_FILES) {
  const fullPath = path.join(ROOT, file);
  const html = fs.readFileSync(fullPath, 'utf8').replace(/<!--[\s\S]*?-->/g, '');
  let match;
  while ((match = ATTR_PATTERN.exec(html)) !== null) {
    const ref = match[1];
    if (!isLocal(ref)) continue;
    const cleanRef = ref.split(/[?#]/)[0];
    if (!cleanRef) continue;
    const target = path.join(ROOT, decodeURIComponent(cleanRef));
    if (!fs.existsSync(target)) {
      missing.push(`${file}: ${ref}`);
    }
  }
}

if (missing.length) {
  console.error('Broken local asset references found:');
  for (const m of missing) console.error(`  - ${m}`);
  process.exit(1);
}

console.log(`OK: all local asset references in ${HTML_FILES.join(', ')} resolve to existing files.`);
